rfcs/taskforce/0001-buildkiteci.md

```diff
@@ -0,0 +1,215 @@
+
+Engineers who are iterating on build rules can run nodes locally or spin up instances to test things using the buildkite CLI
+
+### Description of pipelines
+
+Pipeline are configured with Dhall rather than jinja-templated yaml. Dynamic pipelines support this as a first-class citizen, we do not require running the equivalent "`[test.py](http://test.py)`" before pushing. I have already implemented this on a [side project](https://github.com/bkase/gameboy/blob/master/.buildkite/pipeline.yaml).
+
+**General Pipeline Writing Things**
+
+* Use the [BuildKite monorepo plugin](https://github.com/chronotc/monorepo-diff-buildkite-plugin) for granular test execution.
+* Explicitly opt-in to the dependency graph. Don't support the "wait" instruction that blocks all parallel work.
+
+Assuming this proposal is accepted, more details will follow on how to structure the Dhall descriptions, but here is a rough cut of what it could look like:
+
+```haskell
+-- This is a definition of the package-deb step for Coda
+-- Optional decorations stack on top of the required bits
+-- This step will rerun twice before giving up
+-- It will upload artifacts under /tmp/artifacts/** to google cloud
+-- It depends on coda-build, genesis-ledger, and libp2p-helper steps
+-- Then the 4 required fields are given: a nice label for the UI, the
+--   command to run, a machine-readable key for further dependency graph
+--   manipulation, and a list of globs for granular test execution.
+-- Note: By making dirtyWhen a _required_ field, we set ourselves up for
+--       a world in which CI doesn't run on docs changes
+let packageDebStep =
+   decorate3
+	   (withFlaky 2)
+	   (withArtifacts "/tmp/artifacts/**")
+		 (withDeps [ "coda-build", "genesis-ledger", "libp2p-helper" ])
+	   Command::{
+	     label = ":box: Package deb",
+	     command = "make deb"
+	     key = "package-deb",
+	     dirtyWhen = [ "src/**" ]
+	   }
+```
+
+### Workflow of adding/changing/debugging pipelines
+
+Run buildkite agents locally or on manually spun up compute instances to test rules by targeting specific agents with tags and using the CLI [https://github.com/buildkite/cli](https://github.com/buildkite/cli)
+
+Liberal use of artifact uploading/downloading to keep build re-playable from different steps
+
+## This design solves the above issues
+
+P1 → self-hosted on GKE
+
+P2.1.1 → Parallel by default; opt-in to dependencies
+
+P2.1.2 → Rely on artifacts in gcloud which should be very quick to upload and download from within Gcloud infrastructure
+
+P2.2 → We can choose whatever machines we want
+
+P2.3 → We use dynamic pipelines to expand work based on files touched
+
+P3 → Manual adhoc template expansion is replaced with *automatic* expansion using Dhall
+
+P4 → BuildKite agents can be installed and tagged locally (or on your own instances) to iterate on builds on your own hardware before putting it into production
+
+P5 → Flakiness is explicit in the Dhall configuration
```

Comment:
Not totally sure how I feel about adding a "flake factor" to our tests. Seems to me like we should be identifying why the tests flake in the first place as a standard practice from now on instead. Consider times when there has been a "flake" that we ignored and it turned out there was a race condition or inconsistent bug that was actually causing the flake.

We can consider this as a short term way to support our existing test suite, but we should have the long term goal to be to fix the flakiness altogether (which is captured by the work Team C is doing on the integration test framework, I believe).

Reply:
> Seems to me like we should be identifying why the tests flake in the first place as a standard practice from now on instead

In a Team C meeting we talked about a world where "flaky" means it randomly fails 1 in 20 times, rather than "it fails 2 out of 3 times". "2 out of 3 times" should be insta-remove from the testing suite and a critical issue assigned to someone (if we care about fixing it). But we can follow up with a proposal for flake management there.

> We can consider this as a short term way to support our existing test suite

At least useful for this

Reply:
Ok, sounds good. That definition of "flaky" sounds better than what I had in my head. We should stick that definition of "flaky" in this RFC I think.

Reply:
I'd rather define flaky in a space where there's room to discuss the process and iterate on that definition then block this RFC on that conversation since whatever we come up with there will be expressible if you can specify the # of retries on a job.